fixed integration-provider UX, issue #3210

Lucky-Lodhi2004 opened the pull request:
set integration-provider command now gives error if it doesn't connects to ida or binja, also reverts value to none. Do give me feedback if anything lacks in this PR. i am newbie here

Lucky-Lodhi2004 committed:
fixed integration-provider

disconnect3d commented:
> set integration-provider command now gives error if it doesn't connects to ida or binja, also reverts value to none. Do give me feedback if anything lacks in this PR. i am newbie here

Hi, thanks for the PR! Did u test this code against integration provider? Which ones?

Lucky-Lodhi2004 commented:
> > set integration-provider command now gives error if it doesn't connects to ida or binja, also reverts value to none. Do give me feedback if anything lacks in this PR. i am newbie here
> 
> Hi, thanks for the PR! Did u test this code against integration provider? Which ones?

i ran lint.sh and test.sh, linting gave a lot of errors, errors in the files which i didn't even edited and i'm still puzzled why? as i'm newbie here. when i made PR it ran a lot of tests, all tests passed except lint one and i'm not able to fix it. will you please help me?? This PR can't be merged without passing of lint test. i'm not able to figure out what to do.

disconnect3d commented:
> i ran lint.sh and test.sh, linting gave a lot of errors, errors in the files which i didn't even edited and i'm still puzzled why? as i'm newbie here. when i made PR it ran a lot of tests, all tests passed except lint one and i'm not able to fix it. will you please help me?? This PR can't be merged without passing of lint test. i'm not able to figure out what to do.

You can see the lint errors here: https://github.com/pwndbg/pwndbg/actions/runs/17356838960/job/49271073516?pr=3284

You can typically fix some of them locally with `./lint.sh -f`. Not sure about "getting errors in files you didn't edit" - please show output if you got a different one than on the link above. (Seems like maybe some incorrect setup).

Regarding testing: did you actually run Pwndbg and hit the issue described in #3210? Do your changes fix this specific issue? (I hope you are not trying to fix something without actually trying it out/running it in practice?).

I don't think we have a testcase for this exact scenario. You could create one if you want, though I am not sure how it would look like.

Lucky-Lodhi2004 commented:
> > i ran lint.sh and test.sh, linting gave a lot of errors, errors in the files which i didn't even edited and i'm still puzzled why? as i'm newbie here. when i made PR it ran a lot of tests, all tests passed except lint one and i'm not able to fix it. will you please help me?? This PR can't be merged without passing of lint test. i'm not able to figure out what to do.
> 
> You can see the lint errors here: https://github.com/pwndbg/pwndbg/actions/runs/17356838960/job/49271073516?pr=3284
> 
> You can typically fix some of them locally with `./lint.sh -f`. Not sure about "getting errors in files you didn't edit" - please show output if you got a different one than on the link above. (Seems like maybe some incorrect setup).
> 
> Regarding testing: did you actually run Pwndbg and hit the issue described in #3210? Do your changes fix this specific issue? (I hope you are not trying to fix something without actually trying it out/running it in practice?).
> 
> I don't think we have a testcase for this exact scenario. You could create one if you want, though I am not sure how it would look like.

ok, here's how i set up this project on my ubuntu machine.

first i cloned it and opened on vscodium, nix is installed system wide on my pc so then i ran "nix develop", then i ran setup-dev.sh, then started gdb and then pwndbg prompt opened.

i have free versions of ida and binja and these are not connecting with gdb-pwngdb via plugins so i fixed integration provider feature without them.

i checked "set integration-provider ida" command and it worked properly, ie gave error and reverted value to none.

<img width="760" height="385" alt="Screenshot from 2025-09-06 02-32-05" src="https://github.com/user-attachments/assets/58bb785d-9187-4c36-af2e-3dccc6678b9a" />

<img width="785" height="100" alt="Screenshot from 2025-09-06 02-32-43" src="https://github.com/user-attachments/assets/2c4fb102-e208-45e3-ae59-6909bf08e048" />

after updating the code, i ran ./lint.sh and then this happened.

<img width="698" height="91" alt="Screenshot from 2025-09-06 02-18-15" src="https://github.com/user-attachments/assets/61eb3899-11e9-4075-aa07-27185dbd04bc" />

it did the same for ruff, shfmt and mypy. seems like nix develop didn't help downloading all packages.

<img width="934" height="226" alt="Screenshot from 2025-09-06 02-14-58" src="https://github.com/user-attachments/assets/45c4ea80-edb7-46cb-91e8-82f91682b2f4" />

so i downloaded ruff, shfmt and mypy manually through pip-venv.

<img width="724" height="156" alt="Screenshot from 2025-09-06 02-27-23" src="https://github.com/user-attachments/assets/b518ddf2-b028-4538-b912-fd3626305246" />

and then i ran ./lint.sh and it gave whole lot of errors (errors in the files i didn't even touched.

<img width="829" height="181" alt="Screenshot from 2025-09-06 02-28-18" src="https://github.com/user-attachments/assets/530726b5-579b-426c-a007-c086c96e23de" />
<img width="842" height="535" alt="Screenshot from 2025-09-06 02-28-39" src="https://github.com/user-attachments/assets/04129a35-1af6-4200-9e50-60faad92abb1" />
<img width="842" height="535" alt="Screenshot from 2025-09-06 02-28-49" src="https://github.com/user-attachments/assets/aa46abe5-b3d2-4073-aa55-242b7a35c62a" />
<img width="842" height="535" alt="Screenshot from 2025-09-06 02-29-00" src="https://github.com/user-attachments/assets/33c1a8f1-9c23-4105-b750-1723d0efab3d" />
<img width="842" height="535" alt="Screenshot from 2025-09-06 02-29-10" src="https://github.com/user-attachments/assets/03bf6d11-0de6-4bdf-8837-8ced3d6df4b6" />
<img width="932" height="532" alt="Screenshot from 2025-09-06 02-29-21" src="https://github.com/user-attachments/assets/4ad505cc-6487-494f-b586-55323e068176" />
<img width="932" height="532" alt="Screenshot from 2025-09-06 02-29-28" src="https://github.com/user-attachments/assets/d384ee13-40cc-4719-b0c5-e070693aa490" />
<img width="932" height="532" alt="Screenshot from 2025-09-06 02-29-36" src="https://github.com/user-attachments/assets/c9164e6d-b842-441d-811f-02ce0ba9a33b" />
<img width="932" height="532" alt="Screenshot from 2025-09-06 02-29-42" src="https://github.com/user-attachments/assets/d5ddd753-32de-4161-8740-751b232f68d5" />
<img width="932" height="532" alt="Screenshot from 2025-09-06 02-29-51" src="https://github.com/user-attachments/assets/e684494c-cfab-43d9-aab4-ceda0eca62be" />
<img width="932" height="532" alt="Screenshot from 2025-09-06 02-30-00" src="https://github.com/user-attachments/assets/62e6a6e2-daca-4464-8eed-692a8184ef13" />
<img width="932" height="532" alt="Screenshot from 2025-09-06 02-30-10" src="https://github.com/user-attachments/assets/0272d28f-a428-4386-96c6-8ce5340476d2" />
<img width="932" height="532" alt="Screenshot from 2025-09-06 02-30-20" src="https://github.com/user-attachments/assets/48e49b06-8025-4d6d-a9da-38b0febff8c1" />

and tests.sh showed this

<img width="591" height="163" alt="Screenshot from 2025-09-06 02-31-06" src="https://github.com/user-attachments/assets/94bae3dd-de29-4de8-998d-d3efb4a265ff" />

as the set integration command was working perfectly so i made PR on this repository. it passed all tests but failed lint test like shown on my PR.

<img width="832" height="375" alt="Screenshot from 2025-09-06 02-48-58" src="https://github.com/user-attachments/assets/2f9ee5c9-a2b9-40b4-b0cf-becdb2262a5c" />

<img width="832" height="375" alt="Screenshot from 2025-09-06 02-49-21" src="https://github.com/user-attachments/assets/4ff8648f-dd2e-45ed-9367-264a3302dff7" />


so my questions are:-

1) Am i setting up this project on vscodium wrong way? if yes, then kindly show me correct way to do it

2) can we connect gdb-pwndbg with free versions of ida and binja? if yes, then help me do it

3) Am i missing some modules to download (like uv)?

4) why lint.sh giving so many errors?

kindly answer my above questions and help me setting up this project locally on machine. Also help me merging this PR.

disconnect3d commented:
> so i downloaded ruff, shfmt and mypy manually through pip-venv.

> 4. why lint.sh giving so many errors?

You installed the latest versions of those tools, while we expect specific versions of them. A newer version may pull off more or other warnings etc. Also not sure if the way you use the tools respect our configs for them (which I believe are in pyproject.toml).

> 1. Am i setting up this project on vscodium wrong way? if yes, then kindly show me correct way to do it

I don't use vscodium.

> 2. can we connect gdb-pwndbg with free versions of ida and binja? if yes, then help me do it

No idea, but u can simulate a server with `nc -vvv -l -p 1234` and see if Pwndbg attempts to connect to it or not etc.

k4lizen commented:
the lint script should be using uv to invoke the lint tools, you shouldn't need to install them. looking at your command output, it would seem that `$UV_RUN_LINT` is not set (and it should be). that var is set in `./scripts/common.sh`. Is the environment variable `PWNDBG_NO_UV` set for you somehow? Edit: oh I see you're using nix, I guess it's because of that. you may try to follow the dev environment that does not leverage nix https://pwndbg.re/pwndbg/latest/contributing/setup-pwndbg-dev/

> can we connect gdb-pwndbg with free versions of ida and binja? if yes, then help me do it

Honestly not sure. The integration instructions are here: https://pwndbg.re/pwndbg/latest/tutorials/decompiler-integration/binja-integration/ . I think it's likely you need the non-free versions, but you may check and I will update the docs accordingly

disconnect3d commented on pwndbg/integration/__init__.py:
```diff
@@ -137,16 +156,29 @@ def get_stack_var_name(self, addr: int) -> str | None:
+        import pwndbg.integration.binja
+        if not pwndbg.integration.binja.available():
+            print(message.warn("Binary Ninja not available, reverting to none"))
+            provider_name.value = "none"
```
In one place we do `provider_name.value = "none"` and in another we do `provider_name.set("none")`. What is the difference?

Lucky-Lodhi2004 replied:
should I use 

`provider_name.value = "none"`

 at both places?

disconnect3d commented on pwndbg/integration/__init__.py:
```diff
@@ -91,6 +91,25 @@ def get_stack_var_name(self, addr: int) -> str | None:
+    Ensure the selected integration provider is actually reachable.
+    If not, revert to 'none' and print a notice.
+    """
+    provider = provider_name.value
```
I don't get it. You added this function but it is never used...?

Lucky-Lodhi2004 replied:
I think I should remove this function. 
but I need your help setting up this project on my local machine because it has both nix script and uv script. Should I set up using uv only or nix only or should I use both? Just guide me through the steps after cloning this repo cause It uses specific versions of dependencies.

k4lizen replied:
Just use `uv`, in other words just do
```bash
git clone https://github.com/pwndbg/pwndbg
cd pwndbg
./setup.sh
./setup-dev.sh
```
as per the setup instructions https://pwndbg.re/pwndbg/latest/contributing/setup-pwndbg-dev/ . This will create a venv for you, you don't need to do it yourself. Did you read them fully and found them confusing?

disconnect3d reviewed (changes requested):
See comments.
